src/v/storage/types.h

```diff
@@ -93,7 +103,17 @@ class snapshotable_stm {
  */
 class stm_manager {
 public:
-    void add_stm(ss::shared_ptr<snapshotable_stm> stm) { _stms.push_back(stm); }
+    template<
+      typename T,
+      typename = std::enable_if_t<std::is_base_of_v<snapshotable_stm, T>>>
+    void add_stm(ss::shared_ptr<T> stm) {
+        using type = std::decay_t<T>;
+        _stms.push_back(stm);
+        if constexpr (std::is_same_v<type, compactable_stm>) {
```

Comment:
is `is_same` what we want here? it looks to me like `type` is going to be part of the type hierarchy not necessary exactly compatble_stm?

Reply:
Yeah, I removed this code, but I noticed it should have been, what, `is_base_of<>`?

Reply:
yes, probably. still, it's pretty rare to need such a thing, and is a red flag here that the interface / types can be improved.